### ISSUE TITLE
Allow explicit control of randomness for ensembling

### ISSUE THREAD

**Issue by Jacob-Stevens-Haas** [dynamicslab/pysindy#485]
## Is your feature request related to a problem? Please describe.
`EnsembleOptimizer` uses `_drop_random_samples()` and `np.random.choice` to determine which data/features are involved in each bag.  When running experiments, you can create reproducible ensembles by setting `numpy.random.seed`.  However, when reading experiments, its easy to overlook that this seed is actually used, because it mutates global state, a form of spooky action at a distance. Worse, if someone thinks that it isn't used, deleting `np.random.seed(seed)` won't cause errors.

The solution is to be explicit at least with setting the seed.  [Relevant Jax philosophy](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers).

## Describe the solution you'd like
`EnsembleOptimizer` accepts a `rand_gen: np.random.Generator` object.

## Describe alternatives you've considered
`EnsembleOptimizer` accepts a `seed: int` object.
